pkg/controller/master/addon/addon_processing.go

```diff
@@ -0,0 +1,301 @@
+		return h.processEnablingAddon(aObj)
+	case harvesterv1.AddonDeployed:
+		// addon is enabled, wait and watch for upgrades
+		return h.processAddonUpgrades(aObj)
+	}
+	return aObj, nil
+}
+
+// processDisabledAddon will temporarily set the addon status to AddonDisabling
+// it will then trigger deletion of helm chart and accordingly setup conditions
+// to ensure that job status is tracked to completed
+// when delete job is complete, the addon status is updated to AddonDisabled
+// by default a newly created Addon will be placed into AddonDisabled state
+func (h *Handler) processDisabledAddons(a *harvesterv1.Addon) (*harvesterv1.Addon, error) {
+	aObj := a.DeepCopy()
+
+	if aObj.Status.Status == harvesterv1.AddonDisabled {
+		return aObj, nil
+	}
+
+	// if status is not AddonDisabling and an operation is in progress
+	// then wait for last operation to be completed
+	if aObj.Status.Status != harvesterv1.AddonDisabling && harvesterv1.AddonOperationInProgress.IsTrue(a) {
+		return h.enqueueAfter(aObj)
+	}
+
+	// if status is not AddonDisabling then set status and process object
+	if aObj.Status.Status != harvesterv1.AddonDisabling {
+		aObj.Status.Status = harvesterv1.AddonDisabling
+		return h.addon.UpdateStatus(aObj)
+	}
+
+	return h.waitForAddonDisable(aObj)
+}
+
+func (h *Handler) waitForAddonDisable(a *harvesterv1.Addon) (*harvesterv1.Addon, error) {
+
+	aObj := a.DeepCopy()
+	// get HelmChart and check for processing state
+	hc, owned, err := h.getAddonHelmChart(aObj)
+	if err != nil {
+		return aObj, err
+	}
+
+	// if chart is not owned by AddonController.
+	// disable Addon and place error.
+	if !owned {
+		// no helm chart exists.
+		// no further action required and exit processing of addon
+		if hc == nil {
+			markCompletedCondition(aObj)
+			aObj.Status.Status = harvesterv1.AddonDisabled
+			return h.addon.UpdateStatus(aObj)
+		}
+
+		// if condition already exists, ignore chart
+		// as no further action is possible without manual intervention
+		if harvesterv1.AddonOperationFailed.IsTrue(aObj) {
+			// keep requeing object, as when chart is eventually deleted, this condition will not be met
+			// and chart will be removed
+			return h.enqueueAfter(aObj)
+		}
+
+		// set error condition for addon. no further processing is now possible
+		markErrorCondition(aObj, fmt.Errorf("helmChart %s exists and is now owned by Addon %s. cannot proceed further", hc.Name, aObj.Name))
+		return h.addon.UpdateStatus(aObj)
+	}
+
+	// if helmChart exists trigger deletion of the same
+	if hc.DeletionTimestamp.IsZero() {
+		harvesterv1.AddonOperationInProgress.True(aObj)
+		err := h.helm.Delete(hc.Namespace, hc.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			return aObj, fmt.Errorf("error deleting helm chart: %v", err)
+		}
+		return h.addon.UpdateStatus(aObj)
+	}
+
+	// chart exists and helmChart deletion has been triggered
+	// wait for chart job to be completed
+	if !hc.DeletionTimestamp.IsZero() {
```

Comment:
I think the `!hc.DeletionTimestamp.IsZero()` shouldn't need since the above exists `hc.DeletionTimestamp.IsZero()`.